src/main/java/org/broadinstitute/hellbender/tools/dataflow/transforms/bqsr/BaseRecalibratorFn.java

```diff
@@ -81,167 +59,118 @@
-     * arguments in "toolArgs" and the input file header in "header". The command-line parser will print
-     * any diagnostic messages to "out". If the command-line requested the program not be run
-     * (e.g. it was "--help"), then this method returns null.
-     */
-    public static BaseRecalibratorWorker fromCommandLine(SAMFileHeader header, String commandLine, PrintStream out) {
-        BaseRecalibratorWorker br = new BaseRecalibratorWorker(header);
-        br.BRAC = new BaseRecalibrationArgumentCollection();
-        boolean shouldRun = new CommandLineParser(br).parseArguments(out, Utils.escapeExpressions(commandLine));
-        if (!shouldRun) return null;
-        return br;
-    }
-
-    public static BaseRecalibratorWorker fromArgs(SAMFileHeader header, BaseRecalibrationArgumentCollection toolArgs) {
-        return new BaseRecalibratorWorker(header, toolArgs);
+    public BaseRecalibratorFn(PCollectionView<SAMFileHeader> headerView, PCollectionView<SAMSequenceDictionary> referenceSequenceDictionaryView, BaseRecalibrationArgumentCollection BRAC) {
+        super("BaseRecalibratorFn");
+        this.headerView = headerView;
+        this.referenceSequenceDictionaryView = referenceSequenceDictionaryView;
+        this.BRAC = BRAC;
     }
 
-    private BaseRecalibratorWorker(SAMFileHeader samHeader) {
-        this.samHeader = samHeader;
+    private BaseRecalibratorFn(SAMFileHeader header, BaseRecalibrationArgumentCollection BRAC) {
+        super("BaseRecalibratorFn");
+        this.header = header;
+        this.BRAC = BRAC;
     }
 
-    private BaseRecalibratorWorker(SAMFileHeader samHeader, BaseRecalibrationArgumentCollection toolArgs) {
-        this.samHeader = samHeader;
-        this.BRAC = toolArgs;
+    // saves to BRAC.BRAC.RAC.RECAL_TABLE_FILE.getName()
+    public static void SaveTextualReport(File output, SAMFileHeader header, RecalibrationTables rt, BaseRecalibrationArgumentCollection BRAC) throws IOException {
+        QuantizationInfo qi = new QuantizationInfo(rt, BRAC.RAC.QUANTIZING_LEVELS);
+        BaseRecalibratorFn worker = new BaseRecalibratorFn(header, BRAC);
+        worker.onTraversalStart();
+        worker.saveReport(output, rt, qi, worker.covariates);
     }
 
-    /**
-     * Throws if knownSites is missing.
-     */
-    public void checkClientArguments() {
-        if (null == BRAC.RAC.knownSites || BRAC.RAC.knownSites.size() == 0) {
-            throw new UserException.CommandLineException(NO_DBSNP_EXCEPTION);
-        }
+    @Override
+    public void startBundle(Context c) throws Exception {
+        super.startBundle(c);
+        // can't set things up here because we don't have our side input yet.
+        // So instead we're doing it in processElement.
+        nReadsProcessed = 0;
+        // resetting firstInBundle in case we have multiple bundles via the same instance.
+        firstInBundle = true;
```

Comment:
The code's being a little defensive by not assuming that there's a single bundle per DoFn. The [documentation](https://cloud.google.com/dataflow/java-sdk/JavaDoc/com/google/cloud/dataflow/sdk/transforms/ParDo) suggests that is guaranteed, so if you really want the code to be a few lines shorter we can remove this.